server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java

```diff
@@ -769,474 +832,503 @@ public void start() {
+                        String custom = remoteReadResult.getComponentName().split(CUSTOM_DELIMITER)[1];
+                        clusterStateBuilder.putCustom(custom, (ClusterState.Custom) remoteReadResult.getObj());
+                    }
+                    break;
+                default:
+                    throw new IllegalStateException("Unknown component: " + remoteReadResult.getComponent());
             }
-        } catch (IOException e) {
-            throw new IllegalStateException(
-                String.format(Locale.ROOT, "Error while downloading Settings Metadata - %s", settingsMetadataFileName),
-                e
-            );
+        });
+
+        readIndexRoutingTableResults.forEach(indexRoutingTable -> {
+            indicesRouting.put(indexRoutingTable.getIndex().getName(), indexRoutingTable);
+        });
+
+        metadataBuilder.indices(indexMetadataMap);
+        if (readDiscoveryNodes) {
+            clusterStateBuilder.nodes(discoveryNodesBuilder.get().localNodeId(localNodeId));
         }
+        return clusterStateBuilder.metadata(metadataBuilder)
+            .version(manifest.getStateVersion())
+            .stateUUID(manifest.getStateUUID())
+            .routingTable(new RoutingTable(manifest.getRoutingTableVersion(), indicesRouting))
+            .build();
     }
 
-    private TemplatesMetadata getTemplatesMetadata(String clusterName, String clusterUUID, String templatesMetadataFileName) {
-        try {
-            // Fetch Templates metadata
-            if (templatesMetadataFileName != null) {
-                String[] splitPath = templatesMetadataFileName.split("/");
-                return TEMPLATES_METADATA_FORMAT.read(
-                    globalMetadataContainer(clusterName, clusterUUID),
-                    splitPath[splitPath.length - 1],
-                    blobStoreRepository.getNamedXContentRegistry()
-                );
-            } else {
-                return TemplatesMetadata.EMPTY_METADATA;
+    public ClusterState getClusterStateForManifest(String clusterName, ClusterMetadataManifest manifest, String localNodeId, boolean includeEphemeral)
+        throws IOException {
+        return readClusterStateInParallel(
+            ClusterState.builder(new ClusterName(clusterName)).build(),
+            manifest,
+            manifest.getClusterUUID(),
+            localNodeId,
+            manifest.getIndices(),
+            manifest.getCustomMetadataMap(),
+            manifest.getCoordinationMetadata() != null,
+            manifest.getSettingsMetadata() != null,
+            manifest.getTransientSettingsMetadata() != null,
+            manifest.getTemplatesMetadata() != null,
+            includeEphemeral && manifest.getDiscoveryNodesMetadata() != null,
+            includeEphemeral && manifest.getClusterBlocksMetadata() != null,
+            includeEphemeral ? manifest.getIndicesRouting() : Collections.emptyList(),
```

Comment:
For my understanding, what does this flag `includeEphemeral` mean?

Reply:
@shiv0408  @soosinha  can take this question better.

Reply:
For publication of cluster state, we need ephemeral objects such as DiscoveryNodes, ClusterBlocks. Which are not required for just recovering the cluster state. This flag is set to true while publication and false while the restore flow is true.

Reply:
Got it.